jadx-core/src/main/java/jadx/core/dex/visitors/ModVisitor.java

```diff
@@ -155,6 +159,22 @@ private static void replaceStep(MethodNode mth, InsnRemover remover) {
 						}
 						break;
 
+					case CAST:
+						if (i != 0) {
+							InsnNode pi = block.getInstructions().get(i - 1);
+							if (insn.getResult().getType() == ArgType.BYTE
+									&& pi.getResult().getType() == ArgType.BOOLEAN) {
+								InsnArg zero = new LiteralArg(0, ArgType.BYTE);
+								InsnArg one = new LiteralArg(1, ArgType.BYTE);
+								IfNode ifNode = new IfNode(IfOp.NE, -1, InsnArg.wrapArg(pi), LiteralArg.FALSE);
+								IfCondition condition = IfCondition.fromIfNode(ifNode);
+								TernaryInsn ternary = new TernaryInsn(condition, insn.getResult(), one, zero);
+								replaceInsn(block, i, ternary);
+								remover.addAndUnbind(pi);
+							}
+						}
+						break;
```

Comment:
I don't know why you need previous instruction, try to limit modification scope to minimum.
So I can suggest do this:
```java
					case CAST:
						// replace boolean to byte cast with ternary
						if (insn.getResult().getType() == ArgType.BYTE
								&& insn.getArg(0).getType() == ArgType.BOOLEAN) {
							IfNode ifNode = new IfNode(IfOp.EQ, -1, insn.getArg(0), LiteralArg.TRUE);
							IfCondition condition = IfCondition.fromIfNode(ifNode);
							InsnArg zero = InsnArg.lit(0, ArgType.BYTE);
							InsnArg one = InsnArg.lit(1, ArgType.BYTE);
							TernaryInsn ternary = new TernaryInsn(condition, insn.getResult(), one, zero);
							replaceInsn(block, i, ternary);
						}
						break;
```

Reply:
Sure, this is much better, I guess I was distracted by the "previous Instruction" in the previous code location.

Reply:
Also, should we just use `.isPrimitive()`, instead of the `BYTE||CHAR||SHORT||LONG||DOUBLE||FLOAT`, or even remove it completely, since no casting can be done except to these things (`VOID||BOOLEAN` are the remaining in `primitive`).

Reply:
You can use  `.isPrimitive()` I think it is ok.
But for me looks like you overdid this ))
Yeah sure, you can write these weird conversions in smali but I don't think we can find all of them in real APKs.
Anyway I think this is good work, thanks!